text/0029-block-streaming.md

```diff
@@ -0,0 +1,239 @@
+* Feature Name: `block_streaming`
+* Start Date: 2022-03-01
+* MCIP PR: [mobilecoinfoundation/mcips #29](https://github.com/mobilecoinfoundation/mcips/pull/29)
+* Tracking Issue: [mobilecoin #1433](https://github.com/mobilecoinfoundation/mobilecoin/issues/1433)
+
+# Summary
+[summary]: #summary
+
+This MCIP formalizes the API for getting blocks from consensus nodes.
+
+This includes defining an API for streaming blocks using a dedicated fanout
+mechanism, rather than repeatedly polling S3 for new blocks.
+
+# Motivation
+[motivation]: #motivation
+
+About 40% of the end-to-end transaction time is spent writing to and reading
+from S3. This can be optimized using a dedicated
+[publish/subscribe](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern)
+mechanism to distribute updates across nodes.
+
+<details>
+<summary>Details on the 40% figure</summary>
+
+As shown in the screenshot below (dated 2021-11-17), of the 5 seconds to
+externalize a transaction, we spent almost one second uploading to S3, and
+another second downloading from S3.
+
+![trace screenshot](../images/29-trace.png)
+</details>
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This is a backend implementation change that should be transparent to everyone
+else.
+
+When a node starts up, it will subscribe to updates from its peer consensus
+nodes, and request (or be configured with) the base URL for fetching archived
+blocks, and fetch those in parallel.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There are two mechanisms for getting blocks from consensus nodes:
+1. A streaming API for getting the latest blocks, and
+2. A backfill/discovery API specifying where to find archived blocks.
+
+Each consensus node's ledger is distributed via a local "side-car" process,
+`ledger-distribution`, to an associated directory in the `mobilecoin.chain` S3
```

Comment:
MainNet archive is at `ledger.mainnet.mobilecoin.com` bucket, TestNet is at `mobilecoin.chain` bucket